zipkin-storage/elasticsearch/src/test/java/zipkin2/elasticsearch/integration/ITEnsureIndexTemplate.java

```diff
@@ -0,0 +1,129 @@
+import static zipkin2.TestObjects.DAY;
+import static zipkin2.TestObjects.TODAY;
+
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
+abstract class ITEnsureIndexTemplate {
+
+  ElasticsearchStorage storage;
+
+  /**
+   * Returns a new {@link ElasticsearchStorage.Builder} for connecting to the backend for the test.
+   */
+  protected abstract ElasticsearchStorage.Builder newStorageBuilder(TestInfo testInfo);
+
+  @Test void createZipkinIndexTemplate_getTraces_returnsSuccess(TestInfo testInfo) throws IOException {
+    ElasticsearchStorage.Builder builder = newStorageBuilder(testInfo);
+    storage = builder.build();
+
+    try {
+      /**
+       * Delete all index templates in order to create the "catch-all" index template, because
+       * ES does not allow multiple index templates of the same index_patterns and priority
+       */
+      deleteIndexTemplate("*");
+      setUpCatchAllTemplate();
+
+      /** Create index template with {@link ElasticsearchStorage#ensureIndexTemplates()} */
+      CheckResult check = storage.check();
+
+      assertThat(check.ok()).isTrue();
+
+      Span span = Span.newBuilder().traceId(CLIENT_SPAN.traceId())
+        .id("1")
+        .timestamp(TODAY * 1000L)
+        .putTag("queryTest", "ok")
+        .build();
+
+      storage.spanConsumer().accept(asList(span)).execute();
+
+      assertThat(storage.spanStore().getTraces(QueryRequest.newBuilder()
+        .endTs(TODAY + DAY)
+        .lookback(DAY * 2)
+        .limit(10)
+        .parseAnnotationQuery("queryTest=" + span.tags().get("queryTest"))
+        .build()).execute())
+      .flatExtracting(t -> t).containsExactly(span);
+    }
+    finally {
+      /**
+       * Delete "catch-all" index template so it does not interfere with any other test
+       */
+      deleteIndexTemplate("catch-all");
+      storage.close();
+    }
+  }
+
+  /**
+   * Create a "catch-all" index template with the lowest priority prior to running tests to
+   * ensure that the index templates created during tests with higher priority function as
+   * designed. Only applicable for ES >= 7.8
+   */
+  void setUpCatchAllTemplate() throws IOException {
+    AggregatedHttpRequest updateTemplate = AggregatedHttpRequest.of(
+      RequestHeaders.of(
+        HttpMethod.PUT, catchAllIndexPath(), HttpHeaderNames.CONTENT_TYPE, MediaType.JSON_UTF_8),
+      HttpData.ofUtf8(catchAllTemplate()));
+    Internal.instance.http(storage).newCall(updateTemplate, (parser, contentString) -> null,
+      "update-template").execute();
+  }
+
+  String catchAllIndexPath() {
+    return "/_index_template/catch-all";
+  }
+
+  String catchAllTemplate() {
+    return "{\n"
+      + "  \"index_patterns\" : [\"*\"],\n"
+      + "  \"priority\" : 0,\n"
+      + "  \"template\": {\n"
+      + "    \"settings\" : {\n"
+      + "      \"number_of_shards\" : 1\n"
+      + "    },\n"
+      + "    \"mappings\" : {\n"
+      + "      \"_source\" : { \"enabled\" : true }\n"
```

Comment:
if this is the catch all change you want to make, it would be useful to test that this applies. The test could pass for a subtle reason, then in ES 7.10 it might break and we wouldn't know what it was trying to do either.

For this reason, making a template that sets default which I think this does, maybe isn't a good choice. 
Instead try something with semantic value and also relevant in tracing? ex
```
  "mappings": {
    "properties": {
      "tags.http.status_code": {
        "norms": false,
        "type": "keyword"
      }
    }
  }
```

In this case, your test can switch from making a fake tag "queryTest"->"ok"  to a tag very commonly requested  like "http.status_code" "404"

Then, you can verify your catch-all worked by using HTTP to do the following (except in java):

```
curl -s 'localhost:9200/zipkin-span-*/_search?q=tags.http.status_code:404'
```

Ex what you are doing is literally verifying automatically this https://gist.github.com/adriancole/1af1259102e7a2da1b3c9103565165d7

Reply:
I don't think the suggested secondary template works anymore, given that ES no longer allows multiple matching index templates of the same priority. ES won't merge templates based on order anymore, and will only select 1 with the highest priority. Multiple index template with matching index patterns and priorities will result in error on PUT of the second index template. So in this case, it's either Zipkin's template or catch-all template, and I can't sneak in tags to zipkin-span-* like that.

I do see what you mean, though. And if the objective is to test the catch-all, I think I have to use a new index that doesn't match Zipkin's template at all. Thought?

Reply:
interesting.. thanks for the info! yeah a comment should say this is the case now ( I remember hints of this in the past from @xeraa ). I do think the default template should meddle with ours in a way that would be obvious then. For example, you could change something we don't set? then read back that it is unset (as opposed to inherited/merged) would that do it? 

meanwhile, using http in your test is still probably better as it is more realistic data, but yeah I can now expect people who setup multiple templates to have to do an offline merge of them instead of my gist now!

Reply:
1. There are the [legacy / v1](https://www.elastic.co/guide/en/elasticsearch/reference/7.9/indices-templates-v1.html) templates that can be merged, but I would be careful (multiple templates with the same order will result in a non-deterministic merging order, merging valid templates might lead to an invalid mapping). Right now they are deprecated and we'll see when we'll remove them — you have at least until 8.0.
2. The new [component based approach](https://www.elastic.co/guide/en/elasticsearch/reference/7.9/index-templates.html) still allows you to combine multiple pieces, but you don't merge them (to avoid the two problems mentioned above). This feature was only added in 7.8, so I'm not sure what will be the sanest approach for you — support it in 7.x already or switch over for any (upcoming) 8.x cluster.

Reply:
ok so legacy/v1 would "for some time" retain the current behavior. I suppose in this case this means people in-place upgrading continue to work as this logic only applies when the templates don't already exist. cc also @basvanbeek for perspective on your schema tool. This chat will likely outgrow this PR :P

@ccharnkij for you and to not cage you to this decision tree, I would make a comment to the effect that @xeraa made in point 1. in your unit test and possibly also revise the javadoc/README if doesn't already say this. Basically, this relies on that we are not using legacy/v1 templates, so it is a win, not a merge.

hence we expect them to not merge. So, as mentioned earlier to show that actually happened (sanity check), have your template override something we don't and then magically notice that after the storage component initializes, whatever that setting is, is now ignored.